ADFineGrainedPasswordPolicy: New resource

fl0p3k opened the pull request:
<!--
    Thanks for submitting a Pull Request (PR) to this project.
    Your contribution to this project is greatly appreciated!

    Please prefix the PR title with the resource name,
    e.g. 'ResourceName: My short description'.
    If this is a breaking change, then also prefix the PR title
    with 'BREAKING CHANGE:',
    e.g. 'BREAKING CHANGE: ResourceName: My short description'.

    You may remove this comment block, and the other comment blocks, but please
    keep the headers and the task list.
-->
#### Pull Request (PR) description

This PR adds the `ADFineGrainedPasswordPolicy` resource to the module.

#### This Pull Request (PR) fixes the following issues

- Fixes #584

#### Task list
<!--
    To aid community reviewers in reviewing and merging your PR, please take
    the time to run through the below checklist and make sure your PR has
    everything updated as required.

    Change to [x] for each task in the task list that applies to your PR.
    For those task that don't apply to you PR, leave those as is.
-->
- [x] Added an entry to the change log under the Unreleased section of the
      file CHANGELOG.md. Entry should say what was changed and how that
      affects users (if applicable), and reference the issue being resolved
      (if applicable).
- [x] Resource documentation added/updated in README.md.
- [x] Resource parameter descriptions added/updated in README.md, schema.mof
      and comment-based help.
- [x] Comment-based help added/updated.
- [x] Localization strings added/updated in all localization files as appropriate.
- [x] Examples appropriately added/updated.
- [x] Unit tests added/updated. See [DSC Community Testing Guidelines](https://dsccommunity.org/guidelines/testing-guidelines).
- [x] Integration tests added/updated (where possible). See [DSC Community Testing Guidelines](https://dsccommunity.org/guidelines/testing-guidelines).
- [x] New/changed code adheres to [DSC Community Style Guidelines](https://dsccommunity.org/styleguidelines).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dsccommunity/activedirectorydsc/585)
<!-- Reviewable:end -->

johlju commented:
@fl0p3k can you please look at the failing tests?

johlju commented:
Please see @X-Guardian's https://github.com/dsccommunity/ActiveDirectoryDsc/issues/584#issuecomment-610797852

fl0p3k commented:
I am going to push the unit tests today once completed coding for it

X-Guardian commented:
Ian, please can you complete the new resource proposal issue as requested before continuing with this PR, so that we can discuss the properties and implementation of this resource.

fl0p3k commented:
> Ian, please can you complete the new resource proposal issue as requested before continuing with this PR, so that we can discuss the properties and implementation of this resource.

I have just updated the proposal #584

fl0p3k commented:
> @fl0p3k can you please look at the failing tests?

I corrected the failing tests

fl0p3k commented:
I will be looking over the comments and will provide some updates.

From: Johan Ljunggren<mailto:notifications@github.com>
Sent: Friday, April 10, 2020 10:07 PM
To: dsccommunity/ActiveDirectoryDsc<mailto:ActiveDirectoryDsc@noreply.github.com>
Cc: Ian Florek<mailto:kerolf@live.com>; Mention<mailto:mention@noreply.github.com>
Subject: Re: [dsccommunity/ActiveDirectoryDsc] New ADFineGrainedPasswordPolicy (#585)


@johlju requested changes on this pull request.

Reviewable<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Freviewable.io%2Freviews%2Fdsccommunity%2Factivedirectorydsc%2F585&data=02%7C01%7C%7C8d52a2a26ef84f404c2d08d7ddd643f8%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637221784632958476&sdata=XnUGaKbLv55c5TurrSqno90xw363fyE%2B3AiXZncTfMc%3D&reserved=0> status: 0 of 8 files reviewed, 1 unresolved discussion (waiting on @fl0p3k<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Ffl0p3k&data=02%7C01%7C%7C8d52a2a26ef84f404c2d08d7ddd643f8%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637221784632958476&sdata=RJJDO5UTfN75gAt63zKkk3Docxsl8RUrHmP0X7k0eBo%3D&reserved=0>)

a discussion<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Freviewable.io%2Freviews%2Fdsccommunity%2Factivedirectorydsc%2F585%23-M4bcKXZ9E1jzLi9GSiU%3A-M4bcKXZ9E1jzLi9GSiV%3Ab-jk87ci&data=02%7C01%7C%7C8d52a2a26ef84f404c2d08d7ddd643f8%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637221784632968469&sdata=iAZdOL8EHD3y8noJrAX9TWYaAbS9unRJrouR8p0E2F0%3D&reserved=0> (no related file):
Please resolve the comments in the issue, or discuss them if you have other thoughts.


—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdsccommunity%2FActiveDirectoryDsc%2Fpull%2F585%23pullrequestreview-391737056&data=02%7C01%7C%7C8d52a2a26ef84f404c2d08d7ddd643f8%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637221784632968469&sdata=q0rvvkrRMfi4RaFNbGtCV5uYlmOPCAw5g9QUYOG3dgY%3D&reserved=0>, or unsubscribe<https://nam11.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAMLGSIUUAICNZKLRXQN7G2LRL73J5ANCNFSM4MDSUH7Q&data=02%7C01%7C%7C8d52a2a26ef84f404c2d08d7ddd643f8%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637221784632968469&sdata=v61THQMWNeRl%2BHEcS%2B%2FObD%2F%2Fzj7ZGw%2BvVDSUsLxve6U%3D&reserved=0>.

fl0p3k commented:
*[a discussion](https://reviewable.io/reviews/dsccommunity/activedirectorydsc/585#-M4bcKXZ9E1jzLi9GSiU:-M57iJmc1FzXawIUMQa7:bifattt) (no related file):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Please resolve the comments in the issue, or discuss them if you have other thoughts.
</blockquote></details>

I have made all changes from the review feedback.  The only remaining pending action would be to understand the changes needed to support the PsDscRunAsCredential modifications.  Right now all is working and not sure I need to change anything regarding that.

---


<!-- Sent from Reviewable.io -->

X-Guardian commented:
Ok, as I said in the resource proposal issue, the `ADDomainDefaultPasswordPolicy` resource was written four years ago, and does not use a modern logic pattern. It also not a resource that uses an 'Ensure' property, so the logic flow through the functions is very different.
A better example to follow is the `ADManagedServiceAccount` which has a modern logic flow of an 'Ensure' property resource both in the module and in its unit tests.

With regard to @johlju comments about the `Credential` parameter, I'm happy to leave this in, as it along with the `DomainController` parameter are needed to execute the resource on a non-domain joined machine, where the PsDscRunAsCredential parameter with domain credentials can't be used.

Key features that I would like the new resource to use:
- `Get-TargetResource` function only includes mandatory parameters and parameters explicitly required for the function (DomainController and Credential)
- `Get-TargetResource` function only returns properties defined in the schema.
- No need to use the imutablePropertyMap
- Try/Catch blocks with managed exceptions around every AD Cmdlet call.
- Use `Compare-ResourcePropertyState` common function in the `Test` and `Set` functions.
- Use the `Assert-MemberParameters` common function in the `Test` and `Set` functions to verify the  parameter combination given is valid.
- Examples in the relevant source/Examples/Resources folder with valid script headers.
- Property descriptions as specified in the resource proposal.
- Keep all line lengths to a maximum of 120 characters
- 100% unit test code coverage.

If you are not sure about anything, please ask before coding!

fl0p3k commented:
@X-Guardian 
I've been away from this project for a bit, but I have implemented some changes based on your feedback.  I am hoping that I am meeting the bar at this point to move forward.  My only doubt was how to determine the code coverage the unit tests were evaluating.  I did my best to ensure that all scenarios would get tested.  Thanks.

X-Guardian commented:
Hi @fl0p3k, If you go to the 'Checks' tab on this PR and then select the top check (`dsccommunity.ActiveDirectoryDsc`) then "View more details on Azure Pipelines" then select "Code Coverage", you will see the Code Coverage report for the module. If you then drill down into the `MSFT_ADFineGrainedPasswordPolicy` module you will get a line-by-line report of coverage.

Taking a quick look at your updated code, we are getting closer, but there are still some of the features that I listed that haven't been implemented:

- `Get-TargetResource` function only includes mandatory parameters and parameters explicitly required for the function (DomainController and Credential).
- `Get-TargetResource` function only returns properties defined in the schema.
- Use `Compare-ResourcePropertyState` common function in the Test and Set functions. Currently only used in the Test function.
- Use the `Assert-MemberParameters` common function in the Test and Set functions to verify the parameter combination given is valid.
- Examples in the relevant source/Examples/Resources folder with valid script headers.
- Keep all line lengths to a maximum of 120 characters.
- 100% unit test code coverage.

fl0p3k commented:
> * `Get-TargetResource` function only includes mandatory parameters and parameters explicitly required for the function (DomainController and Credential).
> * `Get-TargetResource` function only returns properties defined in the schema.

@X-Guardian 
Ok, the two items above I am not sure I am on the same page.  I updated the Get-TargetResource function parameters that no longer use the DomainController or Credential parameters.  Also, I updated the Schema to include the SubjectsDifferent read only boolean output.

X-Guardian commented:
Hi @fl0p3k , `Get-TargetResource` needs the `DomainController` and `Credential` parameters to pass to the `Get-ADFineGrainedPasswordPolicy` cmdlet, but it doesn't need the `Subjects`, `SubjectsToInclude` and `SubjectsToExclude` parameters. There is also no need for the `SubjectsDifferent` read only property, which we didn't agree to in the resource proposal. Please stick to that, or discuss first if you think you need to change it.

fl0p3k commented:
> Hi @fl0p3k , `Get-TargetResource` needs the `DomainController` and `Credential` parameters to pass to the `Get-ADFineGrainedPasswordPolicy` cmdlet, but it doesn't need the `Subjects`, `SubjectsToInclude` and `SubjectsToExclude` parameters. There is also no need for the `SubjectsDifferent` read only property, which we didn't agree to in the resource proposal. Please stick to the proposal, or discuss first if you think you need to change it.

@X-Guardian, is there a forum we can chat in realtime (best time as well for you, I'm in PST)? I've been limited being able to focus on this but I have a few days ahead to address updates, but I would like to have a discussion about specific implements before making any more changes.

X-Guardian commented:
Hi @fl0p3k , I'm available on the Slack PowerShell channel at http://slack.poshcode.org/

fl0p3k commented:
Hmm, I think I need an invitation, I am getting a message that the link is no longer active.





<!-- Sent from Reviewable.io -->

X-Guardian commented:
Try the link on the home page: https://poshcode.org/

johlju commented:
@fl0p3k here is more information and links (let me know if they don't work ) https://dsccommunity.org/community/contact/

fl0p3k commented:
I believe it is now ready with 100% code coverage and the changes to support the subjects.  Please review at your convenience.  Thanks!





<!-- Sent from Reviewable.io -->

fl0p3k commented:
___
*[source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.psm1, line 64 at r13](https://reviewable.io/reviews/dsccommunity/activedirectorydsc/585#-MAXY9NL16w3VCzZllRe:-MAYOf5S2dmZFiGL-MM4:bso8c24) ([raw file](https://github.com/dsccommunity/activedirectorydsc/blob/af2ac1cab318127959d49d434d10c15f4b0206ec/source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.psm1#L64)):*
<details><summary><i>Previously, X-Guardian (Simon Heather) wrote…</i></summary><blockquote>

Please remove the `Precedence` parameter from $parameters here, then there is no need to remove it from both `$getADFineGrainedPasswordPolicyParams` and `$getADFineGrainedPasswordPolicySubjectParams` further down.
</blockquote></details>

I get an error if I remove it since it is called out as mandatory in the schema.  This is a required parameter for the 'New-ADFineGrainedPasswordPolicy' cmdlet and I must make it mandatory in the schema I think.


<!-- Sent from Reviewable.io -->

fl0p3k commented:
___
*[source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.psm1, line 92 at r13](https://reviewable.io/reviews/dsccommunity/activedirectorydsc/585#-MAXZEeQ7pd2Ixdd_uPE:-MAYPu8cEKJrO_9MEcpR:b-xj0d5m) ([raw file](https://github.com/dsccommunity/activedirectorydsc/blob/af2ac1cab318127959d49d434d10c15f4b0206ec/source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.psm1#L92)):*
<details><summary><i>Previously, X-Guardian (Simon Heather) wrote…</i></summary><blockquote>

I only get an `ADIdentityNotFoundException` when the password policy doesn't exist, not a lack of subjects. I don't think you need this specific catch as it would have been caught in the previous catch.
</blockquote></details>

I added this incase something else goes wrong (like auth or server error from the parameters).


<!-- Sent from Reviewable.io -->

fl0p3k commented:
___
*[source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.psm1, line 318 at r13](https://reviewable.io/reviews/dsccommunity/activedirectorydsc/585#-MAXaDug7aQaGejO1MB4:-MAYR01T-q6NojkFxr1S:bzdujxp) ([raw file](https://github.com/dsccommunity/activedirectorydsc/blob/af2ac1cab318127959d49d434d10c15f4b0206ec/source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.psm1#L318)):*
<details><summary><i>Previously, X-Guardian (Simon Heather) wrote…</i></summary><blockquote>

What is this for?
</blockquote></details>

I mentioned this in the chat, but when the subjects are null from the Get-TargetResource function, the Compare-ResoucePropertyState does not detect anything mismatched.


<!-- Sent from Reviewable.io -->

fl0p3k commented:
___
*[source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.psm1, line 318 at r13](https://reviewable.io/reviews/dsccommunity/activedirectorydsc/585#-MAXaDug7aQaGejO1MB4:-MAjQsfw1s8YWq2y2hsl:b-ufa0xj) ([raw file](https://github.com/dsccommunity/activedirectorydsc/blob/af2ac1cab318127959d49d434d10c15f4b0206ec/source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.psm1#L318)):*
<details><summary><i>Previously, fl0p3k (Ian Florek) wrote…</i></summary><blockquote>

I have always returned an empty array when the subjects are null.  I made some specific changes to ensure only an empty array object is returned, but I get this:

VERBOSE: [IFLABDC01]:                            [[ADFineGrainedPasswordPolicy]UserAccounts] Evaluating the state of the property 'Subjects'. (ADCOMMON0003)
Cannot bind argument to parameter 'ReferenceObject' because it is null.
    + CategoryInfo          : InvalidData: (:) [], CimException
    + FullyQualifiedErrorId : ParameterArgumentValidationErrorNullNotAllowed,Microsoft.PowerShell.Commands.CompareObjectCommand
    + PSComputerName        : IFLABDC01

VERBOSE: [IFLABDC01]:                            [[ADFineGrainedPasswordPolicy]UserAccounts] The parameter 'Subjects' is in desired state. (ADCOMMON0004)
</blockquote></details>

I figured this one out by using [string[]] instead of object array.


<!-- Sent from Reviewable.io -->

fl0p3k commented:
___
*[source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.psm1, line 318 at r13](https://reviewable.io/reviews/dsccommunity/activedirectorydsc/585#-MAXaDug7aQaGejO1MB4:-MAjR2o_0NfJ73QvzJGs:b-896fix) ([raw file](https://github.com/dsccommunity/activedirectorydsc/blob/af2ac1cab318127959d49d434d10c15f4b0206ec/source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.psm1#L318)):*
<details><summary><i>Previously, fl0p3k (Ian Florek) wrote…</i></summary><blockquote>

I figured this one out by using [string[]] instead of object array.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

fl0p3k commented:
___
*[source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.psm1, line 92 at r13](https://reviewable.io/reviews/dsccommunity/activedirectorydsc/585#-MAXZEeQ7pd2Ixdd_uPE:-MB5MT7d5-IlkcK9holg:b-896fix) ([raw file](https://github.com/dsccommunity/activedirectorydsc/blob/af2ac1cab318127959d49d434d10c15f4b0206ec/source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.psm1#L92)):*
<details><summary><i>Previously, X-Guardian (Simon Heather) wrote…</i></summary><blockquote>

Yes that would be better
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

fl0p3k commented:
___
*[source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.psm1, line 92 at r13](https://reviewable.io/reviews/dsccommunity/activedirectorydsc/585#-MAXZEeQ7pd2Ixdd_uPE:-MBeuONP5uwZvKFEJSkc:btqp1dl) ([raw file](https://github.com/dsccommunity/activedirectorydsc/blob/af2ac1cab318127959d49d434d10c15f4b0206ec/source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.psm1#L92)):*
<details><summary><i>Previously, fl0p3k (Ian Florek) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Any update on closing this out and merging?


<!-- Sent from Reviewable.io -->

johlju commented:
@x-guardian have probably been busy with the day job, I think he come around to this as soon as possible.

X-Guardian commented:
Hi @johlju, I've been talking to @fl0p3k about this PR on Slack.

X-Guardian commented on source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.schema.mof:
```diff
@@ -0,0 +1,22 @@
+    [Write, Description("Specifies the ADPrincipal names the policy is to be applied to, overwrites all existing.")] String Subjects[];
+    [Write, Description("Specifies whether the fine-grained password policy should be present or absent. Default value is 'Present'."), ValueMap{"Present", "Absent"}, Values{"Present", "Absent"}] String Ensure;
+    [Write, Description("Specifies whether password complexity is enabled for the password policy.")] Boolean ComplexityEnabled;
+    [Write, Description("Specifies the length of time that an account is locked after the number of failed login attempts exceeds the lockout threshold (timespan minutes).")] String LockoutDuration;
```
```suggestion
    [Write, Description("Specifies the length of time that an account is locked after the number of failed login attempts exceeds the lockout threshold. The lockout duration must be greater than or equal to the lockout observation time for a password policy. The value must be a string representation of a TimeSpan value.")] String LockoutDuration;
```

X-Guardian commented on source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.schema.mof:
```diff
@@ -0,0 +1,22 @@
+    [Write, Description("Specifies whether the fine-grained password policy should be present or absent. Default value is 'Present'."), ValueMap{"Present", "Absent"}, Values{"Present", "Absent"}] String Ensure;
+    [Write, Description("Specifies whether password complexity is enabled for the password policy.")] Boolean ComplexityEnabled;
+    [Write, Description("Specifies the length of time that an account is locked after the number of failed login attempts exceeds the lockout threshold (timespan minutes).")] String LockoutDuration;
+    [Write, Description("Specifies the maximum time interval between two unsuccessful login attempts before the number of unsuccessful login attempts is reset to 0 (timespan minutes).")] String LockoutObservationWindow;
```
```suggestion
    [Write, Description("Specifies the maximum time interval between two unsuccessful login attempts before the number of unsuccessful login attempts is reset to 0. The lockout observation window must be smaller than or equal to the lockout duration for a password policy. The value must be a string representation of a TimeSpan value.")] String LockoutObservationWindow;
```

X-Guardian commented on source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.schema.mof:
```diff
@@ -0,0 +1,22 @@
+    [Write, Description("Specifies the length of time that an account is locked after the number of failed login attempts exceeds the lockout threshold (timespan minutes).")] String LockoutDuration;
+    [Write, Description("Specifies the maximum time interval between two unsuccessful login attempts before the number of unsuccessful login attempts is reset to 0 (timespan minutes).")] String LockoutObservationWindow;
+    [Write, Description("Specifies the number of unsuccessful login attempts that are permitted before an account is locked out.")] UInt32 LockoutThreshold;
+    [Write, Description("Specifies the minimum length of time before you can change a password (timespan days).")] String MinPasswordAge;
```
```suggestion
    [Write, Description("Specifies the minimum length of time before you can change a password. The value must be a string representation of a TimeSpan value.")] String MinPasswordAge;
```

X-Guardian commented on source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.schema.mof:
```diff
@@ -0,0 +1,22 @@
+    [Write, Description("Specifies the maximum time interval between two unsuccessful login attempts before the number of unsuccessful login attempts is reset to 0 (timespan minutes).")] String LockoutObservationWindow;
+    [Write, Description("Specifies the number of unsuccessful login attempts that are permitted before an account is locked out.")] UInt32 LockoutThreshold;
+    [Write, Description("Specifies the minimum length of time before you can change a password (timespan days).")] String MinPasswordAge;
+    [Write, Description("Specifies the maximum length of time that you can have the same password (timespan days).")] String MaxPasswordAge;
```
```suggestion
    [Write, Description("Specifies the maximum length of time that you can have the same password. The value must be a string representation of a TimeSpan value.")] String MaxPasswordAge;
```

X-Guardian commented on source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/MSFT_ADFineGrainedPasswordPolicy.psm1:
```diff
@@ -0,0 +1,802 @@
+        Specifies whether password complexity is enabled for the password policy.
+
+    .PARAMETER LockoutDuration
+        Specifies the length of time that an account is locked after the number of failed login attempts exceeds the
```
Please adjust the `LockoutDuration`, `LockOutObservationWindow`, `MinPasswordAge` and `MaxPasswordAge` descriptions to the new ones specified in the schema in both the `Test` and `Set` functions.

X-Guardian commented on source/DSCResources/MSFT_ADFineGrainedPasswordPolicy/en-US/about_ADFineGrainedPasswordPolicy.help.txt:
```diff
@@ -0,0 +1,146 @@
+    Write - Boolean
+    Specifies whether password complexity is enabled for the password policy.
+
+.PARAMETER LockoutDuration
```
Please adjust the `LockoutDuration`, `LockOutObservationWindow`, `MinPasswordAge` and `MaxPasswordAge` descriptions to the new ones specified in the schema.

X-Guardian commented:
I just noticed your Unit and Integration tests are in a directory `Tests` with a capital 'T', where they should be in a directory `tests` with a lower case 't'. Windows will merge this into one, so you won't see it on your local PC, but you can see they are different if you look here. https://github.com/fl0p3k/ActiveDirectoryDsc/tree/newADFineGrainedPasswordPolicy. To fix them to lower case, I think you need to delete the files, commit the change, then create them again in the `test` directory with a lowercase 't' then commit them again.

johlju commented:
It is possible to fix by using `git mv`. Use `git ls-files` to find the relative paths that should be changed then use for example:

```
git mv Tests/Unit/test.ps1 tests/Unit/test2.ps1
git mv tests/Unit/test2.ps1 tests/Unit/test.ps1
```

Need to rename the file to a temporary filename then back. After all files have been “moved” then rename the local folder in Windows file system.

X-Guardian commented:
Well done @fl0p3k, we finally got there. Thanks for your patience and perseverance! We have ended up with a good quality resource.

johlju commented:
@fl0p3k really impressive work on this one! Thank you! 😃

fl0p3k commented:
Thanks for the review feedback and patience.  We made a lot of changes and learned a lot. I am grateful to have been a part of the process to get this done.
